pkg/kv/kvserver/replica_lifecycle_datadriven_test.go

```diff
@@ -0,0 +1,340 @@
+}
+
+// replicaInfo contains the basic info about a replica, used for managing its
+// engine (both raft log and state machine) state.
+type replicaInfo struct {
+	id roachpb.FullReplicaID
+	hs raftpb.HardState
+	ts kvserverpb.RaftTruncatedState
+}
+
+// testCtx is a single test's context. It tracks the state of all ranges and any
+// intermediate steps when performing replica lifecycle events.
+type testCtx struct {
+	ranges      map[int]*rangeState
+	nextRangeID int // monotonically-increasing rangeID
+	st          *cluster.Settings
+	// The storage engine corresponds to a single store, (n1, s1).
+	storage storage.Engine
+}
+
+// newTestCtx constructs and returns a new testCtx.
+func newTestCtx() *testCtx {
+	st := cluster.MakeTestingClusterSettings()
+	return &testCtx{
+		ranges:      make(map[int]*rangeState),
+		nextRangeID: 1,
+		st:          st,
+		storage:     storage.NewDefaultInMemForTesting(),
+	}
+}
+
+// close closes the test context's storage engine.
+func (tc *testCtx) close() {
+	tc.storage.Close()
+}
+
+// newRangeState constructs a new rangeState for the supplied descriptor.
+func newRangeState(desc roachpb.RangeDescriptor) *rangeState {
+	return &rangeState{
+		desc:       desc,
+		version:    cluster.MakeTestingClusterSettings().Version.LatestVersion(),
+		nextReplID: len(desc.InternalReplicas) + 1,
+	}
+}
+
+// mustGetRangeState returns the range state for the given range ID.
+func (tc *testCtx) mustGetRangeState(t *testing.T, rangeID int) *rangeState {
+	rs, ok := tc.ranges[rangeID]
+	if !ok {
+		t.Fatalf("range-id %d not found", rangeID)
+	}
+	return rs
+}
+
+func (tc *testCtx) updatePostReplicaCreateState(
+	t *testing.T, ctx context.Context, rs *rangeState, batch storage.Batch,
+) {
+	// Sanity check that we're not overwriting an existing replica.
+	require.Nil(t, rs.replica)
+	sl := stateloader.Make(rs.desc.RangeID)
+	hs, err := sl.LoadHardState(ctx, batch)
+	require.NoError(t, err)
+	ts, err := sl.LoadRaftTruncatedState(ctx, batch)
+	require.NoError(t, err)
+	replID, err := sl.LoadRaftReplicaID(ctx, batch)
+	require.NoError(t, err)
+	rs.replica = &replicaInfo{
+		id: roachpb.FullReplicaID{
+			RangeID:   rs.desc.RangeID,
+			ReplicaID: replID.ReplicaID,
+		},
+		hs: hs,
+		ts: ts,
+	}
+}
+
+func (rs *rangeState) getReplicaDescriptor() *roachpb.ReplicaDescriptor {
+	for i, repl := range rs.desc.InternalReplicas {
+		if repl.NodeID == roachpb.NodeID(1) {
+			return &rs.desc.InternalReplicas[i]
+		}
+	}
+	panic("replica not found")
+}
+
+func (rs *rangeState) String() string {
+	var sb strings.Builder
+	sb.WriteString(fmt.Sprintf("range %d: [%s,%s) replicas=",
+		rs.desc.RangeID, rs.desc.StartKey, rs.desc.EndKey))
+	for i, replica := range rs.desc.InternalReplicas {
+		if i > 0 {
+			sb.WriteString(",")
+		}
+		sb.WriteString(fmt.Sprintf("n%d(r%d)", replica.NodeID, replica.ReplicaID))
+	}
+	if rs.replica != nil {
+		sb.WriteString(fmt.Sprintf("\n		replica (n1/s1): %s\n", rs.replica))
```

Comment:
nit: consider dropping the trailing `\n`.
Rule of thumb: no `\n` in single-line stringers - the caller can add one (or another separator) when they need it, e.g. when combining with other printing.